pkg/kubelet/apis/kubeletconfig/v1beta1/types.go

```diff
@@ -40,298 +40,461 @@ const (
 	// See http://kubernetes.io/docs/user-guide/compute-resources for more detail.
-	KubeReserved map[string]string `json:"kubeReserved"`
-
+	// Default: nil
+	// +optional
+	KubeReserved map[string]string `json:"kubeReserved,omitempty"`
 	// This flag helps kubelet identify absolute name of top level cgroup used to enforce `SystemReserved` compute resource reservation for OS system daemons.
 	// Refer to [Node Allocatable](https://git.k8s.io/community/contributors/design-proposals/node/node-allocatable.md) doc for more information.
+	// Default: ""
+	// +optional
 	SystemReservedCgroup string `json:"systemReservedCgroup,omitempty"`
 	// This flag helps kubelet identify absolute name of top level cgroup used to enforce `KubeReserved` compute resource reservation for Kubernetes node system daemons.
 	// Refer to [Node Allocatable](https://git.k8s.io/community/contributors/design-proposals/node/node-allocatable.md) doc for more information.
+	// Default: ""
+	// +optional
 	KubeReservedCgroup string `json:"kubeReservedCgroup,omitempty"`
 	// This flag specifies the various Node Allocatable enforcements that Kubelet needs to perform.
-	// This flag accepts a list of options. Acceptible options are `pods`, `system-reserved` & `kube-reserved`.
+	// This flag accepts a list of options. Acceptable options are `pods`, `system-reserved` & `kube-reserved`.
 	// Refer to [Node Allocatable](https://git.k8s.io/community/contributors/design-proposals/node/node-allocatable.md) doc for more information.
-	EnforceNodeAllocatable []string `json:"enforceNodeAllocatable"`
-}
-
-type KubeletAuthorizationMode string
-
-const (
-	// KubeletAuthorizationModeAlwaysAllow authorizes all authenticated requests
-	KubeletAuthorizationModeAlwaysAllow KubeletAuthorizationMode = "AlwaysAllow"
-	// KubeletAuthorizationModeWebhook uses the SubjectAccessReview API to determine authorization
-	KubeletAuthorizationModeWebhook KubeletAuthorizationMode = "Webhook"
-)
-
-type KubeletAuthorization struct {
-	// mode is the authorization mode to apply to requests to the kubelet server.
-	// Valid values are AlwaysAllow and Webhook.
-	// Webhook mode uses the SubjectAccessReview API to determine authorization.
-	Mode KubeletAuthorizationMode `json:"mode"`
-
-	// webhook contains settings related to Webhook authorization.
-	Webhook KubeletWebhookAuthorization `json:"webhook"`
-}
-
-type KubeletWebhookAuthorization struct {
-	// cacheAuthorizedTTL is the duration to cache 'authorized' responses from the webhook authorizer.
-	CacheAuthorizedTTL metav1.Duration `json:"cacheAuthorizedTTL"`
-	// cacheUnauthorizedTTL is the duration to cache 'unauthorized' responses from the webhook authorizer.
-	CacheUnauthorizedTTL metav1.Duration `json:"cacheUnauthorizedTTL"`
+	// Default: ["pods"]
+	// +optional
+	EnforceNodeAllocatable []string `json:"enforceNodeAllocatable"` // Do not omitempty, must be able to pass empty list to disable "pods" enforcement.
```

Comment:
this is problematic. would recommend a keyword of "none" instead of an empty list

Reply:
Can you explain why this is problematic?

Reply:
good point, I'll ask @dashpole

Reply:
not all serialization methods distinguish between empty (`[]`) and unset (`nil`). Counting on being able to apply defaults only when this is unset (`nil`) assumes we'll always be able to distinguish between empty and unset when serialized.

Reply:
https://github.com/kubernetes/kubernetes/pull/59515

Reply:
I guess the same problem needs to be solved for EvictionHard too.